third_party/kaldi/src/matrix/kaldi-vector.h

```diff
@@ -0,0 +1,313 @@
+  void MulElements(const VectorBase<Real>& v) {
+    tensor_ *= v.tensor_;
+  }
+
+  // https://github.com/kaldi-asr/kaldi/blob/7fb716aa0f56480af31514c7e362db5c9f787fd4/src/matrix/kaldi-vector.h#L233-L234
+  void Add(Real c) {
+    tensor_ += c;
+  }
+
+  // https://github.com/kaldi-asr/kaldi/blob/7fb716aa0f56480af31514c7e362db5c9f787fd4/src/matrix/kaldi-vector.h#L236-L239
+  void AddVecVec(
+      Real alpha,
+      const VectorBase<Real>& v,
+      const VectorBase<Real>& r,
+      Real beta) {
+    tensor_ = beta * tensor_ + alpha * v.tensor_ * r.tensor_;
+  }
+
+  // https://github.com/kaldi-asr/kaldi/blob/7fb716aa0f56480af31514c7e362db5c9f787fd4/src/matrix/kaldi-vector.h#L246-L247
+  void Scale(Real alpha) {
+    tensor_ *= alpha;
+  }
+
+  // https://github.com/kaldi-asr/kaldi/blob/7fb716aa0f56480af31514c7e362db5c9f787fd4/src/matrix/kaldi-vector.h#L305-L306
+  Real Min() const {
+    if (tensor_.numel()) {
+      return tensor_.min().item().template to<Real>();
+    }
+    return std::numeric_limits<Real>::infinity();
+  }
+
+  // https://github.com/kaldi-asr/kaldi/blob/7fb716aa0f56480af31514c7e362db5c9f787fd4/src/matrix/kaldi-vector.h#L308-L310
+  Real Min(MatrixIndexT* index) const {
+    TORCH_INTERNAL_ASSERT(tensor_.numel());
+    torch::Tensor value, ind;
+    std::tie(value, ind) = tensor_.min(0);
+    *index = ind.item().to<MatrixIndexT>();
+    return value.item().to<Real>();
+  }
+
+  // https://github.com/kaldi-asr/kaldi/blob/7fb716aa0f56480af31514c7e362db5c9f787fd4/src/matrix/kaldi-vector.h#L312-L313
+  Real Sum() const {
+    return tensor_.sum().item().template to<Real>();
+  };
+
+  // https://github.com/kaldi-asr/kaldi/blob/7fb716aa0f56480af31514c7e362db5c9f787fd4/src/matrix/kaldi-vector.h#L320-L321
+  // https://github.com/kaldi-asr/kaldi/blob/7fb716aa0f56480af31514c7e362db5c9f787fd4/src/matrix/kaldi-vector.cc#L718-L736
+  void AddRowSumMat(Real alpha, const MatrixBase<Real>& M, Real beta = 1.0) {
+    Vector<Real> ones(M.NumRows());
+    ones.Set(1.0);
+    this->AddMatVec(alpha, M, kTrans, ones, beta);
+  }
+
+  // https://github.com/kaldi-asr/kaldi/blob/7fb716aa0f56480af31514c7e362db5c9f787fd4/src/matrix/kaldi-vector.h#L323-L324
+  // https://github.com/kaldi-asr/kaldi/blob/7fb716aa0f56480af31514c7e362db5c9f787fd4/src/matrix/kaldi-vector.cc#L738-L757
+  void AddColSumMat(Real alpha, const MatrixBase<Real>& M, Real beta = 1.0) {
+    Vector<Real> ones(M.NumCols());
+    ones.Set(1.0);
+    this->AddMatVec(alpha, M, kNoTrans, ones, beta);
+  }
+
+  // https://github.com/kaldi-asr/kaldi/blob/7fb716aa0f56480af31514c7e362db5c9f787fd4/src/matrix/kaldi-vector.h#L326-L330
+  void AddDiagMat2(
+      Real alpha,
+      const MatrixBase<Real>& M,
+      MatrixTransposeType trans = kNoTrans,
+      Real beta = 1.0) {
+    auto mat = M.tensor_;
+    if (trans == kNoTrans) {
+      tensor_ =
+          beta * tensor_ + torch::diag(torch::mm(mat, mat.transpose(1, 0)));
```

Comment:
If this is called in a tight-loop with small Tensors, you might fare better inlining the operation and using the underlying tensor data pointer instead of calling into torch::mm repeatedly.

Also note that torch::mm does parallelization, but at::parallel_for disables nested parallelism for OpenMP (to avoid oversubscription), which is the default threadpool for PyTorch.